### PR TITLE
KAFKA-12959: Distribute standby and active tasks across threads to better balance load between threads

### PR DESCRIPTION
**Pull request by tim-patterson**
Kafka Streams - Currently while distributing the standby tasks streams does not check if there are threads without any tasks or with less number of tasks. This can lead to few threads getting assigned both active and standby tasks when are threads within the same instance without any tasks assigned.

This PR takes into account active task assignment when assigning standbys to threads to achieve a better balance of tasks across threads

### Committer Checklist (excluded from commit message)
* [ ]  Verify design and implementation
* [ ]  Verify test coverage and CI build status
* [ ]  Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by tim-patterson** [comment]
This contribution is my original work and I license the work to the project under the Apache version 2 licence

**Comment by tim-patterson** [comment]
Hey @ableegoldman I see you've got a few commits on this file, Would you be able to review or point us in the direction of one of your co-committers who might be able to help us here.
(We've been running this in prod for a couple of months now but would prefer not to maintain a fork if we don't have to!)

**Comment by showuon** [comment]
@tim-patterson , thanks for the PR. I'll review the PR next week. Thanks for the improvement.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java
Comment (showuon) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java]: should we keep the `threadLoad` at the end of `assignStatelessTasksToThreads`. We don't need it after `assignStatelessTasksToThreads`, right?

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignorTest.java
Comment (showuon) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignorTest.java]: The `prevTasks00` and `standbyTasks` are meaningless in this test. If we just want to create a mock taskManager, could we use `createMockTaskManager(EMPTY_TASKS, EMPTY_TASKS);` instead?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java
Comment (showuon) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java:814]: could we add comment here to mention this is to combine active stateful assignment + active stateless assignmet...?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java
Comment (showuon) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java]: `threadLoad` is not changed before the end of the `assignTasksToThreads` method. Could we get the value:  `threadLoad.getOrDefault(consumer, 0);` at the beginning of the consumers loop?

Comment (showuon) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java]: Also, to compute the `threadTaskCount`, we always need to add `threadLoad.getOrDefault(consumer, 0)` each time. Could we just make the `minTasksPerThread = minTasksPerThread - threadLoad.getOrDefault(consumer, 0)`, and add comments to it to make it clear, so that we don't need to do this adding `threadLoad.getOrDefault(consumer, 0)` each time? WDYT?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java
Comment (showuon) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java]: Where does the `min + 1` case come from? Could you elaborate more?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java
Comment (showuon) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java:1124]: I think the original algorithm assumes that when reaching this step, all consumers are at the min capacity. So that it could just do this to set `comsuersToFill`:
```java
if (!unassignedStatefulTasks.isEmpty())
       consumersToFill.addAll(consumers);
```

But I didn't see the where we change this algorithm. Please let me know where I missed. Thanks.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java
Comment (showuon) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java]: I was thinking we don't need these loops to get the maxThreadLoad and `consumersToFill`. Do you think if we can put `assignStatelessTasksToThreads` at the end of `assignStatefulTasksToThreads`. So that we will have `maxThreadLoad` and `consumersToFill` directly. So in the `assignStatefulTasksToThreads` method signature, we will have one more parameter like: `boolean shouldAssignStatelessTasks`, or `isActiveTasksAssignment`... something like that. WDYT?

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignorTest.java
Comment (showuon) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignorTest.java]: I'm thinking we can just verify the assignment size is `< 2`. We don't care about the exact number, right?

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignorTest.java
Comment (showuon) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignorTest.java]: Could we add some tests to cover the are less than total tasks (active + standby), but we can still distribute them evenly (the test case should be more common for users), ex:
tasks: TASK_0_0, TASK_0_1, TASK_0_2
consumers: 2 in thread1, 3 in thread2
We expected each thread should have 3 tasks, and distribute evenly.

**Review by showuon — COMMENTED** [review]
@tim-patterson , thanks for the patch! The algorithm changes make sense to me. Left some comments. Thanks.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignorTest.java
Comment (showuon) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignorTest.java]: Also, some tests to make sure the `statelessTasks` are also taking the thread load into account. Thanks.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java
Comment (tim-patterson) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java]: It took me a while to figure out again.
Imagine the case where there's 4 threads,
5 active tasks and 2 standby tasks.

After assigning the actives, 1 of the threads already has 2 tasks.
So when in comes to assigning the standbys the minTasksPerThread is calculated again at 1, so before we've even assigned any standbys we already have some threads at min+1.

Thanks for raising this I'll make a comment :)

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java
Comment (tim-patterson) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java]: Agreed

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java
Comment (tim-patterson) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java:1124]: As commented above

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java
Comment (tim-patterson) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java]: Yeah we can certainly try something like that.
Instead of a boolean the method could take `statefulTasks` and `statelessTasks`,  for the first call we could call it with `statelessTasks = Collection.emptySet()`.
I'll get the other stuff fixed first and then push it up as a separate commit

**Comment by tim-patterson** [comment]
Thanks for the review @showuon.
I've pushed up some of the minor stuff, I'll finish off the unit tests and try merging `assignStatefulTasksToThreads` and `assignStatelessTasksToThreads` together tomorrow morning.
Thanks again
Tim

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java
Comment (showuon) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java]: nit: we can use `threadAssignment.size()` to replace the  `threadTaskCount` variable.  Same as below.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java
Comment (showuon) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java:1119]: Thanks for your explanation. I've got it now, because we've change the total tasks count from original `active tasks` into total ones including active tasks. Thanks.

**Review by showuon — COMMENTED** [review]
Thanks for the update. Left some comments. Waiting for your next update. Thanks.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java
Comment (tim-patterson) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java]: Let me know what you think of the latest commit.
Instead of gluing the methods together I just got rid of `assignStatelessTasksToThreads`, the only real difference between the two methods was the stateful one trying to be sticky.
I just passed in a boolean to control the part of the method that does the stickiness stuff.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignorTest.java
Comment (showuon) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignorTest.java:1094]: nice test!

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignorTest.java
Comment (showuon) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignorTest.java:1133]: I ran it and found there are no stateless tasks included in the assignment. Could you check again?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java
Comment (showuon) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java:1149]: Could we skip this threadLoad update for stateless tasks?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java
Comment (showuon) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java]: nit: we can use stream here:
`int totalTasks = threadLoad.values().stream().reduce(tasksToAssign.size(), Integer::sum);`

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java
Comment (showuon) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java]: Should we mention here that for stateless tasks, we just distribute them evenly to all consumers?

**Review by showuon — COMMENTED** [review]
@tim-patterson , thanks for the update. The stateless tasks update LGTM. It is indeed able to re-use the same logic as the stateful tasks, just no prev-owner sticky assignments. Left some comments. Thanks.

**Comment by showuon** [comment]
@cadonna , do you want to have a second pairs of eyes to look at the PR? Thanks.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java
Comment (tim-patterson) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java:1149]: My only worry with that is that to get a proper balance it would rely on the caller to always assign stateless tasks last which might not be clear to the caller.
Happy to stick an if statement around it if you think it's worth it though.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignorTest.java
Comment (tim-patterson) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignorTest.java:1133]: Good catch

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java
Comment (showuon) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java:1149]: Make sense to me. We can keep it as is. Thanks.

**Review by showuon — APPROVED** [review]
LGTM! Thanks for the improvement! I'll wait until next week to see if @cadonna or @ableegoldman have time to have an another review. Thank you.

**Comment by showuon** [comment]
@tim-patterson , could we rename the PR title? It seems that we don't just assign standby tasks to `threads without active tasks`. Please help update it. Thanks.

**Comment by showuon** [comment]
Failed tests are flaky tests that also failed in trunk branch:
```
[Build / JDK 17 and Scala 2.13 / kafka.server.DynamicBrokerReconfigurationTest.testThreadPoolResize()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-11493/9/testReport/junit/kafka.server/DynamicBrokerReconfigurationTest/Build___JDK_17_and_Scala_2_13___testThreadPoolResize___2/)
    [Build / JDK 17 and Scala 2.13 / kafka.server.DynamicBrokerReconfigurationTest.testThreadPoolResize()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-11493/9/testReport/junit/kafka.server/DynamicBrokerReconfigurationTest/Build___JDK_17_and_Scala_2_13___testThreadPoolResize__/)
    [Build / JDK 11 and Scala 2.13 / kafka.server.DynamicBrokerReconfigurationTest.testThreadPoolResize()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-11493/9/testReport/junit/kafka.server/DynamicBrokerReconfigurationTest/Build___JDK_11_and_Scala_2_13___testThreadPoolResize__/)
    [Build / JDK 11 and Scala 2.13 / kafka.server.DynamicBrokerReconfigurationTest.testThreadPoolResize()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-11493/9/testReport/junit/kafka.server/DynamicBrokerReconfigurationTest/Build___JDK_11_and_Scala_2_13___testThreadPoolResize___2/)
```

**Comment by showuon** [comment]
@tim-patterson , thanks for the contribution!

**Pull request merged**
